File: processing/syntactic/expressions/atomic/constant_expr.py
from __future__ import annotations

from processing.codegen.code_builder import CodeBuilder
from processing.lexical.tokens.token import Token
from processing.syntactic.expressions.expr import Expr
from processing.syntactic.expressions.expression_groups import AtomExpression
from processing.text_location import span_marker


class ConstantExpr(AtomExpression):
    """const_expr:
       CONST_TOKEN;
    """

    def __init__(
            self,
            parent: Expr = None,
            literal: Token = None
    ):
        super().__init__(parent)
        self.literal = literal

    @span_marker
    def parse(self) -> ConstantExpr:
        self.literal = self.stream.eat_any()
        return self

    def to_axion(self, c: CodeBuilder):
        c += self.literal

    def to_csharp(self, c: CodeBuilder):
        c += self.literal

File: Axion.Core/Processing/Syntactic/Expressions/Atomic/ConstantExpr.cs
using Axion.Core.Processing.Lexical.Tokens;
using Axion.Core.Processing.Syntactic.Expressions.Common;
using Axion.Core.Processing.Syntactic.Expressions.TypeNames;
using Magnolia.Attributes;
using static Axion.Specification.TokenType;

namespace Axion.Core.Processing.Syntactic.Expressions.Atomic;

/// <summary>
///     <code>
///         const-expr:
///             CONST-TOKEN | STRING+;
///     </code>
/// </summary>
[Branch]
public partial class ConstantExpr : AtomExpr {
    [Leaf] Token? literal;

    public override TypeName? InferredType => Literal?.InferredType;

    public ConstantExpr(Node parent) : base(parent) { }

    public static ConstantExpr True(Node parent) {
        return new(parent) {
            Literal = new Token(parent.Unit, KeywordTrue)
        };
    }

    public static ConstantExpr False(Node parent) {
        return new(parent) {
            Literal = new Token(parent.Unit, KeywordFalse)
        };
    }

    public static ConstantExpr Nil(Node parent) {
        return new(parent) {
            Literal = new Token(parent.Unit, KeywordNil)
        };
    }

    public static ConstantExpr ParseNew(Node parent) {
        return new ConstantExpr(parent).Parse();
    }

    public ConstantExpr Parse() {
        Literal ??= Stream.EatAny();
        return this;
    }
}
